Compatibility updates for Solr 9.8.1

marktriggs opened the pull request:
@demiankatz I've rejiggered things a little for Solr 9.8.1 with the hope of future-proofing a bit while I'm at it.  I've got the unit tests passing, but I don't have a really good VuFind test setup running at the moment.  If things don't look right, let me know and I'll get my environment resurrected :)

------------------
The `IndexSearcher.search(query, collector)` method is now deprecated. We used custom collectors in several places: sometimes just to count numbers of results; other times because we wanted to walk a full result set but didn't care about relevance ranking.

The places using TotalHitCountCollector can be replaced by a call to IndexSearcher.count(), so no problem there.

For the BibDB, which used to use a custom Collector, I've switched it to use a standard call to IndexSearcher(query, limit) and to process the resulting TopDocs.  Using the higher-level API should hopefully be more future-proof.

It does mean we're now firing the query twice (once to get the result count, once to accumulate the results into the right-sized array), and it also means we're now computing relevance ranking scores we don't use.  But the queries involved are so simple that neither of these overheads will be noticeable to end users.

marktriggs committed:
Compatibility updates for Solr 9.8.1

The `IndexSearcher.search(query, collector)` method is now deprecated.
We used custom collectors in several places: sometimes just to count
numbers of results; other times because we wanted to walk a full
result set but didn't care about relevance ranking.

The places using TotalHitCountCollector can be replaced by a call to
IndexSearcher.count(), so no problem there.

For the BibDB, which used to use a custom Collector, I've switched it
to use a standard call to IndexSearcher(query, limit) and to process
the resulting TopDocs.  Using the higher-level API should hopefully be
more future-proof.

It does mean we're now firing the query twice (once to get the result
count, once to accumulate the results into the right-sized array), and
it also means we're now computing relevance ranking scores we don't
use.  But the queries involved are so simple that neither of these
overheads will be noticeable to end users.

marktriggs committed:
Remove unused variables/imports

demiankatz committed:
Merge branch 'dev' into solr-9.8-upgrade

demiankatz commented on src/main/java/org/vufind/util/DeweyCallNormalizer.java:
```diff
@@ -1,7 +1,7 @@
 
 import org.marc4j.callnum.DeweyCallNumber;
-import java.util.logging.Logger;
+// import java.util.logging.Logger;
```
I see that you've commented out logger-related code here and in LCCallNormalizer. Is that intentional? If so, should we just remove this code if it's no longer needed? Commenting it out without explanation might just confuse us in the future.

marktriggs replied:
Yeah, I commented those out because my language server reports the `log` variable as unused and I try not to ignore those warnings, but I'm in the habit of leaving those lines commented because they're somewhat likely to be wanted again at some point, and somewhat annoying to add back from scratch :)

The annoyance is figuring out which of the `Logger` classes is actually the one that will do the logging you want.  Even in this little project, there are 8 different classes called `Logger`, but only one of them will actually work.

demiankatz replied:
Fair enough, I withdraw my complaint. :-)

demiankatz reviewed (changes requested):
Thanks, @marktriggs!

This looks reasonable to me; unfortunately, I'm encountering one failing test when I update my VuFind .jar files with these new versions:

```
1) VuFindTest\Integration\Connection\SolrTest::testAlphaBrowseUseInstead
VuFindSearch\Backend\Exception\RemoteErrorException: 500 Server Error

/usr/local/vufind/module/VuFindSearch/src/VuFindSearch/Backend/Exception/HttpErrorException.php:68
/usr/local/vufind/module/VuFindSearch/src/VuFindSearch/Backend/Solr/Connector.php:521
/usr/local/vufind/module/VuFindSearch/src/VuFindSearch/Backend/Solr/Connector.php:446
/usr/local/vufind/module/VuFindSearch/src/VuFindSearch/Backend/Solr/Connector.php:330
/usr/local/vufind/module/VuFindSearch/src/VuFindSearch/Backend/Solr/Backend.php:418
/usr/local/vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Connection/SolrTest.php:93
```

I was hoping that the Solr logs would tell me more, but I just see this:

![VirtualBox_VuFind (Mainline, PHP 8) Development Box_04_06_2025_15_47_25](https://github.com/user-attachments/assets/d5f079cc-42af-4415-a1c9-3c1678f4214e)

Any ideas? I'm out of time for today but am happy to dig into this further tomorrow if that would save you the trouble of setting up your own environment. Just let me know if you can think of any relevant leads for me to follow up!

Also, see below for a more trivial comment.

marktriggs commented:
@demiankatz Do any errors jump out in your Solr log?

I think I've resurrected my development setup, but I'm not seeing errors when testing through the UI.  I'm running from the `dev` branch on this commit:

```
commit 387dd22e4c03e9710a895931eec6e261d381a18a (HEAD -> dev, origin/dev, origin/HEAD)
Author: Thomas Wagener <wagener@hebis.uni-frankfurt.de>
Date:   Wed Jun 4 19:55:49 2025 +0200

    Remove deprecated static config locator (#4409)
```

Solr is reporting running version `9.8.1`, and I'm using versions of `browse-indexing.jar` and `browse-handler.jar` built from this PR.  And Java 11, if that's relevant.

demiankatz commented:
@marktriggs, here's the error from the Solr log:

```
2025-06-05 01:05:16.035 ERROR (qtp5395829-34-localhost-59) [c: s: r: x:biblio t:localhost-59] o.a.s.s.HttpSolrCall 500 Exception => java.lang.IllegalArgumentException: numHits must be > 0; please use TotalHitCountCollectorManager if you just need the total hit count
	at org.apache.lucene.search.TopScoreDocCollectorManager.<init>(TopScoreDocCollectorManager.java:67)
```

Note that the test that's triggering this is calling the low-level Solr classes directly rather than running through the user interface -- so it may be doing something that's not accessible to the end user, but which is available through internal APIs.

demiankatz commented:
Here is [the offending code](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/tests/integration-tests/src/VuFindTest/Connection/SolrTest.php#L92-L93).

marktriggs committed:
Don't fire a query against the Solr index if we know it won't match

It turns out you can't pass a limit of zero to a TopDocs search.  It
throws an exception if you try.

Add a unit test to catch this case.

marktriggs commented:
Ahh thanks!  So it turns out it's now an error to fire a search at a Lucene index and ask for zero hits back.

I think that's unlikely to happen in real life because we're searching for headings we know to be in the index.  But still, might as well handle it.  I've also added a unit test to the browse code to catch that case.

demiankatz reviewed (approved):
Thanks, @marktriggs, all tests are passing for me now! As always, I really appreciate your help with this issue!

marktriggs commented:
@demiankatz Glad to help :)
